# CS559 Assignment3 Problem1 k-means
Student Name: Chenxu Wang  
CWID: 10457625  
E-mail: cwang101@stevens.edu  

In [1]:
import numpy as np
import math 

In [2]:
# define a function to calculating the Euclidean distance between two data points
def distance(x1, x2):
    e1 = (x1[0]-x2[0])**2
    e2 = (x1[1]-x2[1])**2
    d = math.sqrt(e1+e2)
    return d

In [17]:
# Iterate all data
# Find the cluster that has minimal distance between this point and the center of this cluster
# update the marker of this point
# data: 10*2
# centers: 3*2
# Output:
# markers: 10x3
def update_marker(data, centers):
    # set the dimension of makers up
    s = (10,3)
    markers = np.zeros(s, dtype = int) # 10x3
    for i in range(10):
        d0 = distance(data[i], centers[0])
        # print(d0)
        d1 = distance(data[i], centers[1])
        # print(d1)
        d2 = distance(data[i], centers[2])
        # print(d2)
        items = [(d0, 0), (d1, 1), (d2, 2)]
        dictionary = dict(items)
        max_dis = min(d0, d1, d2)
        j = dictionary[max_dis]
        markers[i][j] = 1
    return markers   

In [4]:
# Minimize J with respect to the center_k, keeping markers_k fixed.
# Mean of all of the data points Xn assigned to cluster k.
# data: 10x2
# markers: 10x3
# Output:
# centers: 3x2
def update_center(data, markers):
    # Find the points in cluster 1
    c0 = []
    for i in range(10):
        if(markers[i][0]== 1):
            c0.append(data[i])
    cluster0 = np.array(c0)
    # update the center of cluster 1
    center0 = np.sum(cluster0, axis = 0) / len(cluster0)

    # Find the points in cluster 2
    c1 = []
    for i in range(10):
        if(markers[i][1]== 1):
            c1.append(data[i])
    cluster1 = np.array(c1)
    # update the center of cluster 2
    center1 = np.sum(cluster1, axis = 0)/ len(cluster1)
    
    # Find the points in cluster 3
    c2 = []
    for i in range(10):
        if(markers[i][2]== 1):
            c2.append(data[i])
    cluster2 = np.array(c2)
    # update the center of cluster 3
    center2 = np.sum(cluster2, axis = 0)/ len(cluster2)
    
    centers = np.array([center0, center1, center2])
    return centers

In [14]:
# For example

data = np.array([[5.9, 3.2], [4.6, 2.9], [6.2, 2.8], [4.7, 3.2], [5.5, 4.2], [5.0, 3.0],
              [4.9, 3.1], [6.7, 3.1], [5.1, 3.8], [6.0, 3.0]])
# initialize center_of_cluster
center_of_cluster = np.array([[6.2, 3.2], [6.6, 3.7], [6.5, 3.0]]) 
# initialize cluster_marker
# set the dimension of makers up
#s = (10,3)
#cluster_marker = np.zeros(s, dtype = int) # 10x3


In [18]:
# Given different initialization of center_of_cluster, cluster_marker
# you will have different array

# But better show outputs like following code.

for i in range(4):
    cluster_marker = update_marker(data, center_of_cluster)
    center_of_cluster = update_center(data, cluster_marker)
    print("iteration : ", i+1)
    print("center of RED is", center_of_cluster[0, :])
    print("center of GREEN is", center_of_cluster[1, :])
    print("center of BLUE is", center_of_cluster[2, :])
    print("--------------------------------------------------")

iteration :  1
center of RED is [4.8  3.05]
center of GREEN is [5.3 4. ]
center of BLUE is [6.2   3.025]
--------------------------------------------------
iteration :  2
center of RED is [4.8  3.05]
center of GREEN is [5.3 4. ]
center of BLUE is [6.2   3.025]
--------------------------------------------------
iteration :  3
center of RED is [4.8  3.05]
center of GREEN is [5.3 4. ]
center of BLUE is [6.2   3.025]
--------------------------------------------------
iteration :  4
center of RED is [4.8  3.05]
center of GREEN is [5.3 4. ]
center of BLUE is [6.2   3.025]
--------------------------------------------------


In [16]:
print(cluster_marker)

[[0 0 1]
 [1 0 0]
 [0 0 1]
 [1 0 0]
 [0 1 0]
 [1 0 0]
 [1 0 0]
 [0 0 1]
 [0 1 0]
 [0 0 1]]


# Problem1
(1) After one iteration, the center of the first cluster is (4.8  3.05).  
(2) After two iteration, the center of the second cluster is (5.3 4.0).  
(3) After two iteration, the center of the second cluster is (6.2   3.025).  
(4) There are 2 iterations required for the clusters to converge.  